File: src/deephaven_mcp/client/_protobuf.py
"""Wrapper classes for protobuf messages used by the Deephaven client module.

This module provides convenience wrappers around protobuf messages to offer
more Pythonic interfaces and utility methods for working with protobuf objects.
These wrappers simplify interaction with the underlying protobuf API by providing
standardized access methods, property accessors, and serialization helpers.

The module contains wrapper classes for various protobuf message types used
in client-server communication, including query configurations, state information,
status enums, and authentication tokens.

Classes:
    ProtobufWrapper: Base class providing common functionality for protobuf wrappers.
    CorePlusQueryStatus: Wrapper for query status enum values with convenience methods.
    CorePlusToken: Wrapper for authentication token messages.
    CorePlusQueryConfig: Wrapper for query configuration messages.
    CorePlusQueryState: Wrapper for query state messages.
    CorePlusQueryInfo: Wrapper for comprehensive query information messages.
"""

from typing import Any, NewType

from google.protobuf.json_format import MessageToDict, MessageToJson
from google.protobuf.message import Message

from ._base import is_enterprise_available

if is_enterprise_available:
    from deephaven_enterprise.client.controller import ControllerClient
else:
    ControllerClient = None


# Type definitions
CorePlusQuerySerial = NewType("CorePlusQuerySerial", int)
"""Type representing the serial number of a query."""


class ProtobufWrapper:
    """A wrapper for a protobuf message that provides convenience methods.

    This base class provides common functionality for all protobuf wrapper classes,
    including dictionary and JSON serialization. It enforces non-null protobuf messages
    and provides a consistent interface for accessing the underlying protobuf object.

    Example:
        >>> wrapper = ProtobufWrapper(pb_message)
        >>> dict_data = wrapper.to_dict()
        >>> json_str = wrapper.to_json()
    """

    def __init__(self, pb: Message):
        if pb is None:
            raise ValueError("Protobuf message cannot be None")

        self._pb = pb

    def __repr__(self) -> str:
        """Returns a string representation of the wrapper."""
        pb_type = type(self._pb).__name__
        return f"<{self.__class__.__name__} wrapping {pb_type}>"

    @property
    def pb(self) -> Message:
        """The underlying protobuf message."""
        return self._pb

    def to_dict(self) -> dict[str, Any]:
        """Returns the protobuf message as a dictionary.

        Returns:
            Dict[str, Any]: A dictionary representation of the protobuf message
        """
        return MessageToDict(self._pb, preserving_proto_field_name=True)

    def to_json(self) -> str:
        """Returns the protobuf message as a JSON string.

        Returns:
            str: A JSON string representation of the protobuf message
        """
        return MessageToJson(self._pb, preserving_proto_field_name=True)


class CorePlusQueryStatus(ProtobufWrapper):
    """Wrapper for a PersistentQueryStatusEnum value.

    This class wraps a protobuf enum value for query status.
    It provides utility methods for checking status conditions
    by delegating to ControllerClient methods.

    This class simplifies status checking with properties like is_running, is_completed,
    is_terminal and is_uninitialized. It also supports string comparison with the status
    name and direct comparison with other status objects.

    Example:
        >>> status = CorePlusQueryStatus(pb_status_enum)
        >>> if status.is_running:
        ...     print(f"Query is running with status: {status}")
        >>> if status == "RUNNING":
        ...     print("Status matches string 'RUNNING'")

    This corresponds to PersistentQueryStatusEnum in the protobuf definition:
    https://docs.deephaven.io/protodoc/20240517/#io.deephaven.proto.controller.PersistentQueryStatusEnum
    """

    def __init__(
        self,
        status: "deephaven_enterprise.proto.controller.PersistentQueryStatusEnum",  # noqa: F821
    ):
        """Initialize with a protobuf status enum value.

        Args:
            status: The protobuf enum value for query status
        """
        super().__init__(status)

    def __str__(self) -> str:
        """Return the string representation of the status."""
        return self.name

    def __eq__(self, other: object) -> bool:
        """Compare this status with another status or string."""
        if isinstance(other, CorePlusQueryStatus):
            return self.pb == other.pb
        elif isinstance(other, str):
            return self.name == other
        return self.pb == other

    @property
    def name(self) -> str:
        """Get the string name of the status."""
        return ControllerClient.status_name(self.pb)

    @property
    def is_running(self) -> bool:
        """Check if the query status is running."""
        return ControllerClient.is_running(self.pb)

    @property
    def is_completed(self) -> bool:
        """Check if the query status is completed."""
        return ControllerClient.is_completed(self.pb)

    @property
    def is_terminal(self) -> bool:
        """Check if the query status is in a terminal state."""
        return ControllerClient.is_terminal(self.pb)

    @property
    def is_uninitialized(self) -> bool:
        """Check if the query status is uninitialized."""
        return ControllerClient.is_status_uninitialized(self.pb)


class CorePlusToken(ProtobufWrapper):
    """
    Wrapper for authentication Token message.

    This class wraps a protobuf Token message (type: deephaven_enterprise.proto.auth_pb2.Token)
    to provide a more convenient interface for accessing token information such as service name,
    issuer, and expiration time.

    It simplifies the interaction with authentication tokens in the Deephaven environment,
    allowing for easier token management and validation. The wrapped token contains
    information about authentication credentials, including:
    - The token value itself (a string used for authentication)
    - Service information (what service the token authenticates with)
    - Expiration time (when the token becomes invalid)
    - Issuer information (who created/issued the token)
    - User identity information (who the token represents)

    Args:
        token: The protobuf Token message to wrap (type: deephaven_enterprise.proto.auth_pb2.Token)

    Example:
        >>> token = CorePlusToken(pb_token)
        >>> token_dict = token.to_dict()
        >>> print(f"Token expires: {token_dict.get('expires_at')}")
        >>> print(f"Token issuer: {token_dict.get('issuer')}")

    This corresponds to Token in the protobuf definition:
    https://docs.deephaven.io/protodoc/20240517/#io.deephaven.proto.auth.Token
    """

    def __init__(
        self, token: "deephaven_enterprise.proto.auth_pb2.Token"  # noqa: F821
    ):
        """Initialize with a protobuf Token message.

        Args:
            token: The protobuf Token message to wrap
        """
        super().__init__(token)


class CorePlusQueryConfig(ProtobufWrapper):
    """Wrapper for a PersistentQueryConfigMessage.

    Provides a more Pythonic interface to the query configuration. This class wraps
    the protobuf configuration message for persistent queries to make it easier to
    work with in Python code.

    The configuration contains settings that determine how a persistent query is executed,
    including but not limited to:
    - Query name and description
    - Memory allocation (heap size)
    - CPU allocation and priority
    - Server/node placement constraints
    - Engine type and version
    - Query source definition (script, table, application)
    - Initialization parameters
    - Timeout settings
    - Replication settings

    Query configurations are typically created using helper methods like
    `make_temporary_config()` from the controller client, rather than constructed manually.

    Example:
        >>> config = CorePlusQueryConfig(pb_config)
        >>> config_dict = config.to_dict()
        >>> print(f"Query name: {config_dict.get('name')}")
        >>> print(f"Heap size: {config_dict.get('heap_size_mb')} MB")
        >>> print(f"Engine: {config_dict.get('engine_type')}")

    This corresponds to PersistentQueryConfigMessage in the protobuf definition:
    https://docs.deephaven.io/protodoc/20240517/#io.deephaven.proto.persistent_query.PersistentQueryConfigMessage
    """

    def __init__(
        self,
        config: "deephaven_enterprise.proto.persistent_query_pb2.PersistentQueryConfigMessage",  # noqa: F821
    ):
        """Initialize with a protobuf PersistentQueryConfigMessage.

        Args:
            config: The protobuf configuration message to wrap
        """
        super().__init__(config)


class CorePlusQueryState(ProtobufWrapper):
    """Wrapper for a PersistentQueryStateMessage.

    This class wraps the protobuf state message for persistent queries to provide
    a more convenient interface for accessing state information such as query status.

    The state contains runtime information about a persistent query, including:
    - Current execution status (running, stopped, failed, etc.)
    - Runtime metrics and resource usage statistics
    - Initialization and update timestamps
    - Execution history and progress information
    - Error information and diagnostics (if applicable)
    - Worker node/host information

    The most commonly accessed property is `status`, which provides a CorePlusQueryStatus
    object that can be used to determine the current execution state of the query.

    Example:
        >>> state = CorePlusQueryState(pb_state)
        >>> status = state.status
        >>> if status.is_running:
        ...     print(f"Query is running")
        >>> elif status.is_terminal:
        ...     print(f"Query has terminated with status: {status}")
        ...     # Access error information if available
        ...     state_dict = state.to_dict()
        ...     if 'error_message' in state_dict:
        ...         print(f"Error: {state_dict['error_message']}")

    This corresponds to PersistentQueryStateMessage in the protobuf definition:
    https://docs.deephaven.io/protodoc/20240517/#io.deephaven.proto.persistent_query.PersistentQueryStateMessage
    """

    def __init__(
        self,
        state: "deephaven_enterprise.proto.persistent_query_pb2.PersistentQueryStateMessage",  # noqa: F821
    ):
        """Initialize with a protobuf PersistentQueryStateMessage.

        Args:
            state: The protobuf state message to wrap
        """
        super().__init__(state)

    @property
    def status(self) -> CorePlusQueryStatus:
        """Returns the status of the query."""
        return CorePlusQueryStatus(self.pb.status)


class CorePlusQueryInfo(ProtobufWrapper):
    """Wrapper for a PersistentQueryInfoMessage.

    Provides a more Pythonic interface to the query info by wrapping the
    nested config and state messages into their respective wrapper classes.

    This is a comprehensive wrapper that combines configuration, state, and replication
    information for a persistent query. It serves as the main access point for working
    with persistent queries and provides convenient access to all aspects of a query's
    definition and runtime state.

    Key components of a CorePlusQueryInfo include:
    - config: The CorePlusQueryConfig containing the query's configuration parameters
    - state: A CorePlusQueryState representing the primary query's current state (may be None)
    - replicas: A list of CorePlusQueryState objects for any replica instances of the query
    - spares: A list of CorePlusQueryState objects for any spare instances of the query

    This class is typically obtained from the controller client's `map()` or `get()`
    methods and provides all the information needed to monitor and manage a query.

    Example:
        >>> info = CorePlusQueryInfo(pb_info)
        >>> # Access query configuration
        >>> config = info.config
        >>> config_dict = config.to_dict()
        >>> print(f"Query name: {config_dict.get('name')}")
        >>>
        >>> # Access query state
        >>> state = info.state
        >>> if state and state.status.is_running:
        ...     print(f"Query is running with {len(info.replicas)} replicas")
        >>>
        >>> # Check replication status
        >>> if info.replicas:
        ...     print(f"Query has {len(info.replicas)} active replicas")
        ...     for i, replica in enumerate(info.replicas):
        ...         print(f"Replica {i} status: {replica.status}")

    This corresponds to PersistentQueryInfoMessage in the protobuf definition:
    https://docs.deephaven.io/protodoc/20240517/#io.deephaven.proto.persistent_query.PersistentQueryInfoMessage
    """

    def __init__(
        self,
        info: "deephaven_enterprise.proto.persistent_query_pb2.PersistentQueryInfoMessage",  # noqa: F821
    ):
        """Initialize with a protobuf PersistentQueryInfoMessage.

        Args:
            info: The protobuf query info message to wrap
        """
        super().__init__(info)
        self._config: CorePlusQueryConfig = CorePlusQueryConfig(info.config)
        self._state: CorePlusQueryState | None = (
            CorePlusQueryState(info.state) if info.state else None
        )
        self._replicas: list[CorePlusQueryState] = [
            CorePlusQueryState(r) for r in info.replicas
        ]
        self._spares: list[CorePlusQueryState] = [
            CorePlusQueryState(s) for s in info.spares
        ]

    @property
    def config(self) -> CorePlusQueryConfig:
        """The wrapped configuration of the query."""
        return self._config

    @property
    def state(self) -> CorePlusQueryState | None:
        """The wrapped state of the query, if present."""
        return self._state

    @property
    def replicas(self) -> list[CorePlusQueryState]:
        """A list of wrapped replica states for the query."""
        return self._replicas

    @property
    def spares(self) -> list[CorePlusQueryState]:
        """A list of wrapped spare states for the query."""
        return self._spares

File: src-ts/client/protobuf.ts
/**
 * Wrapper types for DHE query objects used by the Deephaven client module.
 *
 * In the TypeScript port, the Python protobuf-based wrappers are replaced with
 * JavaScript-native wrappers around the DHE JS API objects. The DHE JS API is
 * loaded at runtime from the DHE server via `@deephaven/jsapi-nodejs`.
 *
 * The architecture follows a consistent pattern where each wrapper class:
 * - Wraps a specific DHE JS API object type
 * - Inherits from the {@link DheWrapper} base class
 * - Provides specialized methods relevant to the wrapped object's domain
 * - Maintains access to the underlying DHE object when needed
 *
 * Classes:
 *   {@link DheWrapper}: Base class providing common functionality for DHE object wrappers.
 *   {@link CorePlusQueryStatus}: Wrapper for query status values with convenience methods.
 *   {@link CorePlusToken}: Wrapper for authentication token objects.
 *   {@link CorePlusQueryConfig}: Wrapper for query configuration objects.
 *   {@link CorePlusQueryState}: Wrapper for query state objects.
 *   {@link CorePlusQueryInfo}: Wrapper for comprehensive query information objects.
 *
 * Type Definitions:
 *   {@link CorePlusQuerySerial}: Type representing the serial number of a query.
 */

/**
 * Maps each PQ state string to its lifecycle category.
 *
 * Categories:
 * - ACTIVE: PQ is processing data; session_id is present (RUNNING, EXECUTING)
 * - TRANSITIONAL: PQ is between stable states; do not branch on a specific value
 * - TERMINAL: state will not change without user action; STOPPED and FAILED can be restarted
 * - INVALID: UNSPECIFIED is a protobuf zero-value sentinel; should not appear at runtime
 */
export const PQ_STATES: Record<string, string> = {
  RUNNING: "ACTIVE",
  EXECUTING: "ACTIVE",
  UNINITIALIZED: "TRANSITIONAL",
  CONNECTING: "TRANSITIONAL",
  AUTHENTICATING: "TRANSITIONAL",
  ACQUIRING_WORKER: "TRANSITIONAL",
  INITIALIZING: "TRANSITIONAL",
  STOPPING: "TRANSITIONAL",
  DISCONNECTED: "TRANSITIONAL",
  STOPPED: "TERMINAL",
  FAILED: "TERMINAL",
  KILLED: "TERMINAL",
  COMPLETED: "TERMINAL",
  ERROR: "TERMINAL",
  UNSPECIFIED: "INVALID",
};

/**
 * Type representing the serial number of a persistent query.
 *
 * A query serial is a unique identifier assigned to each persistent query in the Deephaven system.
 * It is used to reference, lookup, and manage specific query instances. Query serials are
 * numbers assigned incrementally by the controller service when queries are created.
 *
 * Unlike query names (which are optional and user-defined), serials are guaranteed to be unique
 * within a Deephaven server instance and are the primary key for query identification in the API.
 *
 * @example
 * ```typescript
 * const querySerial: CorePlusQuerySerial = 12345 as CorePlusQuerySerial;
 * const info = await controllerClient.get(querySerial);
 * ```
 */
export type CorePlusQuerySerial = number & { readonly _brand: unique symbol };

/**
 * A wrapper for a DHE object that provides convenience methods.
 *
 * This base class provides common functionality for all DHE object wrapper classes.
 * It enforces non-null objects and provides a consistent interface for accessing
 * the underlying DHE object.
 *
 * @example
 * ```typescript
 * const wrapper = new DheWrapper(dheObject);
 * const dict = wrapper.toDict();
 * const json = wrapper.toJson();
 * ```
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE objects have arbitrary shape
export class DheWrapper<T extends Record<string, any> = Record<string, any>> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE objects have arbitrary shape
  protected readonly _pb: T;

  /**
   * Initialize with a DHE object.
   *
   * @param pb - The DHE object to wrap. Must not be null/undefined.
   * @throws {Error} If the provided object is null/undefined.
   */
  constructor(pb: T) {
    if (pb === null || pb === undefined) {
      throw new Error("Protobuf message cannot be None");
    }
    this._pb = pb;
  }

  toString(): string {
    return `<${this.constructor.name} wrapping ${Object.prototype.toString.call(this._pb)}>`;
  }

  /**
   * The underlying DHE object.
   */
  get pb(): T {
    return this._pb;
  }

  /**
   * Return the DHE object as a plain dictionary.
   *
   * @returns A dictionary representation of the DHE object.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- config is an arbitrary object
  toDict(): Record<string, any> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE objects have arbitrary shape
    return { ...(this._pb as Record<string, any>) };
  }

  /**
   * Return the DHE object as a JSON string.
   *
   * @returns A JSON string representation of the DHE object.
   */
  toJson(): string {
    return JSON.stringify(this._pb);
  }
}

/**
 * Wrapper for a query status value providing status checking functionality.
 *
 * This class wraps a query status object from the DHE JS API, which represents
 * the current lifecycle state of a query or worker process in the Deephaven system.
 * It provides utility methods and properties for checking status conditions,
 * simplifying status-based decision making.
 *
 * Common status values include:
 * - UNINITIALIZED: Initial state before query execution begins
 * - INITIALIZING: Query is being set up but not yet running
 * - RUNNING: Query is actively executing and processing data
 * - STOPPING: Query is in the process of shutting down gracefully
 * - STOPPED: Query has been gracefully terminated
 * - COMPLETED: Query has finished execution successfully
 * - FAILED: Query encountered an error and terminated abnormally
 * - KILLED: Query was forcibly terminated
 *
 * @example
 * ```typescript
 * const status = new CorePlusQueryStatus({ name: "RUNNING" });
 * if (status.isRunning) {
 *   console.log(`Query is running with status: ${status}`);
 * }
 * ```
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE status objects have arbitrary shape
export class CorePlusQueryStatus extends DheWrapper<Record<string, any>> {
  /**
   * Initialize with a DHE query status object.
   *
   * @param status - The DHE status object. Must have a `name` property.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE status objects have arbitrary shape
  constructor(status: Record<string, any>) {
    super(status);
  }

  toString(): string {
    return this.name;
  }

  equals(other: CorePlusQueryStatus | string | unknown): boolean {
    if (other instanceof CorePlusQueryStatus) {
      return this.pb === other.pb || this.name === other.name;
    }
    if (typeof other === "string") {
      return this.name === other;
    }
    return this.pb === other;
  }

  /**
   * The string name of the status, with the DHE `PQS_` prefix stripped.
   *
   * The underlying library may return prefixed status names (e.g., `"PQS_RUNNING"`);
   * this property returns the short form (e.g., `"RUNNING"`) for logging,
   * display, and string comparisons.
   *
   * @returns The stripped status name (e.g., `"RUNNING"`, `"COMPLETED"`, `"FAILED"`).
   */
  get name(): string {
    const raw: string = String(this._pb["name"] ?? this._pb["status"] ?? "UNSPECIFIED");
    return raw.startsWith("PQS_") ? raw.slice("PQS_".length) : raw;
  }

  /**
   * Check if the query status is running.
   *
   * A running query is actively processing data and executing its defined operations.
   *
   * @returns `true` if the query is in a running state, `false` otherwise.
   */
  get isRunning(): boolean {
    return PQ_STATES[this.name] === "ACTIVE";
  }

  /**
   * Check if the query status is completed.
   *
   * A completed query has finished its execution successfully.
   *
   * @returns `true` if the query has completed successfully, `false` otherwise.
   */
  get isCompleted(): boolean {
    return this.name === "COMPLETED";
  }

  /**
   * Check if the query status is in a terminal state.
   *
   * Terminal states represent the end of a query's lifecycle. No further state transitions
   * will occur once a query reaches a terminal state.
   *
   * @returns `true` if the query is in a terminal state, `false` otherwise.
   */
  get isTerminal(): boolean {
    return PQ_STATES[this.name] === "TERMINAL";
  }

  /**
   * Check if the query status is uninitialized.
   *
   * @returns `true` if the query is in the uninitialized state, `false` otherwise.
   */
  get isUninitialized(): boolean {
    return this.name === "UNINITIALIZED";
  }

  /**
   * Check if the query status is initializing.
   *
   * @returns `true` if the query is in the initializing state, `false` otherwise.
   */
  get isInitializing(): boolean {
    return this.name === "INITIALIZING";
  }
}

/**
 * Wrapper for authentication token objects in the Deephaven authentication system.
 *
 * This class wraps a DHE JS API token object to provide a more convenient interface
 * for accessing token information such as service name, issuer, and expiration time.
 * Tokens are central to Deephaven's authentication and authorization system.
 *
 * @example
 * ```typescript
 * const token = new CorePlusToken(dheToken);
 * const tokenDict = token.toDict();
 * ```
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE token objects have arbitrary shape
export class CorePlusToken extends DheWrapper<Record<string, any>> {
  /**
   * Initialize with a DHE token object.
   *
   * @param token - The DHE token object to wrap. Must not be null/undefined.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE token objects have arbitrary shape
  constructor(token: Record<string, any>) {
    super(token);
  }
}

/**
 * Wrapper for a query configuration object defining how a query should be executed.
 *
 * Provides a more convenient interface to the query configuration from the DHE JS API.
 *
 * @example
 * ```typescript
 * const config = new CorePlusQueryConfig(dheConfig);
 * const configDict = config.toDict();
 * console.log(`Query name: ${configDict.name}`);
 * ```
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE config objects have arbitrary shape
export class CorePlusQueryConfig extends DheWrapper<Record<string, any>> {
  /**
   * Initialize with a DHE query configuration object.
   *
   * @param config - The DHE configuration object to wrap. Must not be null/undefined.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE config objects have arbitrary shape
  constructor(config: Record<string, any>) {
    super(config);
  }
}

/**
 * Wrapper for a query state object.
 *
 * This class wraps the DHE JS API state object for persistent queries to provide
 * a more convenient interface for accessing state information such as query status.
 *
 * @example
 * ```typescript
 * const state = new CorePlusQueryState(dheState);
 * const status = state.status;
 * if (status.isRunning) {
 *   console.log("Query is running");
 * }
 * ```
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE state objects have arbitrary shape
export class CorePlusQueryState extends DheWrapper<Record<string, any>> {
  /**
   * Initialize with a DHE query state object.
   *
   * @param state - The DHE state object to wrap. Must not be null/undefined.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE state objects have arbitrary shape
  constructor(state: Record<string, any>) {
    super(state);
  }

  /**
   * The query's current lifecycle status.
   *
   * @returns Wrapper for the query's current status value.
   */
  get status(): CorePlusQueryStatus {
    return new CorePlusQueryStatus(this._pb["status"] ?? { name: "UNSPECIFIED" });
  }
}

/**
 * Wrapper for a comprehensive query information object.
 *
 * Provides a more convenient interface to the query info by wrapping the
 * nested config and state objects into their respective wrapper classes.
 *
 * @example
 * ```typescript
 * const info = new CorePlusQueryInfo(dheQueryInfo);
 * const config = info.config;
 * const state = info.state;
 * if (state && state.status.isRunning) {
 *   console.log(`Query is running with ${info.replicas.length} replicas`);
 * }
 * ```
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE query info objects have arbitrary shape
export class CorePlusQueryInfo extends DheWrapper<Record<string, any>> {
  private readonly _config: CorePlusQueryConfig;
  private readonly _state: CorePlusQueryState | undefined;
  private readonly _replicas: CorePlusQueryState[];
  private readonly _spares: CorePlusQueryState[];

  /**
   * Initialize with a DHE query info object.
   *
   * @param info - The DHE query info object to wrap. Must not be null/undefined.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE query info objects have arbitrary shape
  constructor(info: Record<string, any>) {
    super(info);
    this._config = new CorePlusQueryConfig(info["config"] ?? {});
    this._state = info["state"] ? new CorePlusQueryState(info["state"]) : undefined;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE objects have arbitrary shape
    this._replicas = (info["replicas"] ?? []).map((r: Record<string, any>) => new CorePlusQueryState(r));
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- DHE objects have arbitrary shape
    this._spares = (info["spares"] ?? []).map((s: Record<string, any>) => new CorePlusQueryState(s));
  }

  /**
   * The wrapped configuration of the query.
   *
   * @returns Wrapper for the query's configuration settings.
   */
  get config(): CorePlusQueryConfig {
    return this._config;
  }

  /**
   * The wrapped state of the query, if present.
   *
   * @returns Wrapper for the query's primary state information, or `undefined` if no state exists.
   */
  get state(): CorePlusQueryState | undefined {
    return this._state;
  }

  /**
   * A list of wrapped replica states for the query.
   *
   * @returns A list of state wrappers for all active replicas. Empty if no replicas exist.
   */
  get replicas(): CorePlusQueryState[] {
    return this._replicas;
  }

  /**
   * A list of wrapped spare states for the query.
   *
   * @returns A list of state wrappers for all spare instances. Empty if no spares exist.
   */
  get spares(): CorePlusQueryState[] {
    return this._spares;
  }
}

// Backwards-compatible alias: the base class used to be called ProtobufWrapper
export { DheWrapper as ProtobufWrapper };
